### ISSUE TITLE
Supplemental Code Review for GridNode

### ISSUE THREAD

**Issue by samreid** [phetsims/griddle#59]
GridNode was developed in https://github.com/phetsims/scenery-phet/issues/580 and I had a chance to take it for a test drive in Gravity and Orbits for https://github.com/phetsims/gravity-and-orbits/issues/328.  It's mostly working great but I saw some opportunities for improvements, so thought I should share them.

In this code:

```js
  /**
   * Returns an array of positions in model coordinates of lines that are within grid view bounds, modified
   * by transform. This is useful if you need to decorate the GridNode with labels or other things.
   * @public
   *
   * @param {string} lineType
   * @returns {[]}
   */
  getHorizontalLinePositionsInGrid( lineType ) {
```

- [x] `lineType` should have type `LineType`
- [x] the return type should be `{number[]}`

For this code:

```js
  getSpacingFromLineType( lineType ) {
    let lineSpacing = null;
    if ( lineType === LineType.MAJOR_VERTICAL ) {
      lineSpacing = this.majorVerticalLineSpacing;
    }
    else if ( lineType === LineType.MAJOR_HORIZONTAL ) {
      lineSpacing = this.majorHorizontalLineSpacing;
    }
    else if ( lineType === LineType.MINOR_VERTICAL ) {
      lineSpacing = this.minorVerticalLineSpacing;
    }
    else if ( lineType === LineType.MINOR_HORIZONTAL ) {
      lineSpacing = this.minorHorizontalLineSpacing;
    }

    return lineSpacing;
  }
```

- [ ] Should this be managed by a `Map`?

For this code:

```js
    // @private - Paths for each of the sets of lines
    this.minorHorizontalLines = new Path( null, options.minorLineOptions );
    this.minorVerticalLines = new Path( null, options.minorLineOptions );
    this.majorHorizontalLines = new Path( null, options.majorLineOptions );
    this.majorVerticalLines = new Path( null, options.majorLineOptions );
```

- [ ] Did you consider using one Path to draw all minor lines and one Path to draw all major lines?  What is the advantage of separating out the horizontal from vertical?

For this code:

```js
setLineSpacings( majorVerticalLineSpacing, majorHorizontalLineSpacing, minorVerticalLineSpacing, minorHorizontalLineSpacing ) {
```

- [ ] It seems like there should be a mutator for each one, instead of having to supply all 4 if you only want to change 1.

For this code:

```js
      // {Object} - passed to the Path for minor lines
      minorLineOptions: {
        stroke: 'grey',
        lineWidth: 1
      },

      // {null|Property.<ModelViewTransform2>} - model-view transform for the grid for line spacings and other
      // transformations
      modelViewTransformProperty: null,

      // {Object} - passed to the Path for major lines
      majorLineOptions: {
        stroke: 'black',
        lineWidth: 3
      }
```

- [x] It would be nice to put the `minorLineOptions` next to the `majorLineOptions` instead of having `modelViewTransformProperty` in between.

- [ ] Have you considered creating a more general type like `SingleGridNode` that shows vertical and horizontal lines, then a `GridNode` just has `majorGridNode = new SingleGridNode()` and `minorGridNode = new SingleGridNode()`?  Or are there cross-dependencies that make this impossible (such as skipping minor grid lines that are obscured by the major grid lines)?  I don't think there are cross-dependencies, but thought I would ask.  This could even be a private class if we are worried about complexity leaking out, but I could also see that some clients would just want to instantiate a `SingleGridNode` itself.

For this code:

```js
    // check to see if model-view transform flipped relative bottom and top due to an inverse vertical transformation
    if ( modelGridBottom < modelGridTop ) {
      for ( let y = modelGridBottom + distanceToGridLine; y <= modelGridBottom + modelHeight; y += spacing ) {
        positions.push( y );
      }
    }
    else {
      for ( let y = modelGridTop + distanceToGridLine; y <= modelGridTop + modelHeight; y += spacing ) {
        positions.push( y );
      }
    }
```

- [x] Would it be simpler to run a single loop like `for (let y=Math.min(...);y<=Math.max(...);y+=spacing){}`?

For this code:

```js
  getVerticalLinePositionsInGrid( lineType ) {
    assert && assert( lineType === LineType.MAJOR_VERTICAL || lineType === LineType.MINOR_VERTICAL, 'lineType should be one of the vertical line spacings' );

    const spacing = this.getSpacingFromLineType( lineType );
    assert && assert( spacing === null || typeof spacing === 'number', `spacing not defined for ${lineType}` );

    const positions = [];

    // no lines of this line type in grid, return empty array
    if ( spacing === null ) {
      return positions;
    }

    const modelViewTransform = this.modelViewTransformProperty.get();

    const modelGridLeft = modelViewTransform.viewToModelX( 0 );
    const modelWidth = modelViewTransform.viewToModelDeltaX( this.gridWidth );

    // distance from left edge of the gridNode to the first vertical line, in model coordinates
    const remainderToLine = Utils.toFixedNumber( modelGridLeft % spacing, 10 );
    const distanceToGridLine = ( spacing - remainderToLine ) % spacing;

    for ( let x = modelGridLeft + distanceToGridLine; x <= modelGridLeft + modelWidth; x += spacing ) {
      positions.push( x );
    }

    return positions;
  }

  /**
   * Returns an array of positions in model coordinates of lines that are within grid view bounds, modified
   * by transform. This is useful if you need to decorate the GridNode with labels or other things.
   * @public
   *
   * @param {string} lineType
   * @returns {[]}
   */
  getHorizontalLinePositionsInGrid( lineType ) {
    assert && assert( lineType === LineType.MAJOR_HORIZONTAL || lineType === LineType.MINOR_HORIZONTAL, 'lineType should be one of the horizontal line spacings' );

    const spacing = this.getSpacingFromLineType( lineType );
    assert && assert( spacing === null || typeof spacing === 'number', `spacing not defined for ${lineType}` );

    const positions = [];

    // no lines of this LineType in grid, return empty array
    if ( spacing === null ) {
      return positions;
    }

    const modelViewTransform = this.modelViewTransformProperty.get();
    const modelGridBottom = modelViewTransform.viewToModelY( 0 );
    const modelGridTop = modelViewTransform.viewToModelY( this.gridHeight );
    const modelHeight = Math.abs( modelViewTransform.viewToModelDeltaY( this.gridHeight ) );

    // distance from top edge of the gridNode to the first horizontal line
    const remainderToGridLine = Utils.toFixedNumber( modelGridBottom % spacing, 10 );
    const distanceToGridLine = ( spacing - remainderToGridLine ) % spacing;

    // check to see if model-view transform flipped relative bottom and top due to an inverse vertical transformation
    if ( modelGridBottom < modelGridTop ) {
      for ( let y = modelGridBottom + distanceToGridLine; y <= modelGridBottom + modelHeight; y += spacing ) {
        positions.push( y );
      }
    }
    else {
      for ( let y = modelGridTop + distanceToGridLine; y <= modelGridTop + modelHeight; y += spacing ) {
        positions.push( y );
      }
    }

    return positions;
  }
```

- [x] There is a lot of duplication between vertical and horizontal.  Perhaps this would be obviated by other recommendations from this review, or if we keep it perhaps it could be simplified by using `PHET_CORE/Orientation` and just having one implementation (especially paired with the Min/Max strategy in the preceding review comment).

In this code:

```js
    // distance from left edge of the gridNode to the first vertical line, in model coordinates
    const remainderToLine = Utils.toFixedNumber( modelGridLeft % spacing, 10 );
    const distanceToGridLine = ( spacing - remainderToLine ) % spacing;
```

- [x] It looks like it is rounding, can you please add a comment why rounding is useful here?

**Comment by jessegreenberg** [comment]
Thanks for the review comments! I made commits involving the incorrect/lacking documentation above, and addressed the code duplication with `getHorizontalLinePositionsInGrid` and `getVerticalLinePositionsInGrid`.

> Should this be managed by a Map

I played around with it but I think it is a wash. The spacing fields on GridNode  would be passed to the Map by value so the Map would need to be updated every time spacing changed. I kind of like that `getSpacingFromLineType` doesn't require that. But I don't feel strongly.

> Did you consider using one Path to draw all minor lines and one Path to draw all major lines? What is the advantage of separating out the horizontal from vertical?

Yes, it used to be like this. This was changed to use four paths in https://github.com/phetsims/griddle/issues/57 because the Paths use lineDashOffset so that line dashes will appear to transform with the grid. The line dash offset may be different for horizontal and vertical lines. See the griddle demo for an example.

> It seems like there should be a mutator for each one, instead of having to supply all 4 if you only want to change 1.

This API for GridNode was discussed in https://github.com/phetsims/scenery-phet/issues/601. Specifically, it was requested that we enforce major spacing > minor spacing, minor spacing must be a multiple of major spacing, and at least one type of spacing is defined. When there is a setter for each individual spacing, there is ambiguity in the order of using setters. For example, doing this would cause assertion errors:
```js
gridNode.setMinorHorizontalSpacing( 5 );
gridNode.setMajorHorizontalSpacing( 10 );


gridNode.setMinorHorizontalSpacing( 7 ); // 10 is not a multiple of 7 so this transient state would cause assertion errors
gridNode.setMajorHorizontalSpacing( 14 );
```

And so we decided that all four need to be specified at once.

> Have you considered creating a more general type like SingleGridNode that shows vertical and horizontal lines, then a GridNode just has majorGridNode = new SingleGridNode() and minorGridNode = new SingleGridNode()

I had not considered this. I don't see yet how this alternative would improve the API or implementation. Perhaps if someone only needed a `SingleGridNode` they would avoid instantiating the extra Paths and it could be more light-weight. But that alone doesn't seem like enough to change GridNode.

Reassigning to @samreid to review comments.
